Gh plus mhd

moxcodes opened the pull request:
## Proposed changes

Adds executables for combined Generalized Harmonic and Valencia GRMHD systems.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Dependencies

- [x] #3177
- [x] #3190

~Two commits (`Allow different Metavariables in VisitAndReturn functions` and `Template ObserveTimeStep on System rather than Metavariables` are shared with #3131 -- I'll rebase and remove them from whichever PR is merged later)~

moxcodes committed:
Add EvolveGhValenciaDivClean.hpp

moxcodes committed:
Remove static assert for conservative form in Initialization

This action is also usable for 'hybrid' systems like GH+MHD in which some of the
evolved variables are in flux conservative form and others are not.

moxcodes committed:
Allow some nonconservative variables in SetVariables

permits 'hybrid' systems like GH+MHD in which some variables are in
flux conservative form and some are not.

moxcodes committed:
Add EvolveGhValenciaDivCleanWithHorizon.hpp

nilsdeppe commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
```diff
@@ -0,0 +1,499 @@
+
+  using time_stepper_tag = Tags::TimeStepper<
+      std::conditional_t<local_time_stepping, LtsTimeStepper, TimeStepper>>;
+  // using analytic_solution_fields = typename system::variables_tag::tags_list;
```
can this be removed?

nilsdeppe commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
```diff
@@ -0,0 +1,499 @@
+  // using analytic_solution_fields = typename system::variables_tag::tags_list;
+
+  struct ObservationType {};
+  using element_observation_type = ObservationType;
```
what is this used for?

nilsdeppe commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/System.hpp:
```diff
@@ -47,13 +47,24 @@ struct System {
-  using gradient_tags = gradient_variables;
+  using gradients_tags = gradient_variables;
   using sourced_variables = typename grmhd_system::sourced_variables;
+  // This causes some extra derivatives of GRMHD quantities to be calculated
```
could you say where the derivatives are computed? I thought during the evolution only derivatives of the `gradient_variables` are computed, so why are the GRMHD variables listed there? Though I assume the `grmhd_system::gradient_variables` is empty, so now I'm really confused what this comment is telling me :P

moxcodes replied:
Ah, yep, I had misunderstood the `DerivCompute` tag added in `TimeStepperHistory` -- I think you're right that there's no redundancy because the `grmhd_system::gradient_variables` is appropriately an empty list

moxcodes committed:
remove unused `ObservationType` from GH metavars

moxcodes commented:
Rebased on develop, no longer dependent.
Also posted fixups for the comments, and added another commit to remove the unused `ObservationType` from the GH executable as well.
Cheers!

nilsdeppe commented:
LGTM, please go ahead and squash :)

moxcodes commented:
fixups squashed in.
Cheers!

moxcodes commented:
(had to squash in updates to the test input files that changed)

nilsdeppe commented:
@moxcodes looks like something went wrong since CI is very unhappy

moxcodes commented:
yep, sorry, there were also changes to the horizon finder that required another rebase and some tweaks.

kidder commented on src/Evolution/Systems/RelativisticEuler/Valencia/System.hpp:
```diff
@@ -40,6 +40,7 @@ struct System {
       tmpl::list<Tags::TildeD, Tags::TildeTau, Tags::TildeS<Dim>>>;
   using flux_variables =
       tmpl::list<Tags::TildeD, Tags::TildeTau, Tags::TildeS<Dim>>;
+  using non_conservative_variables = tmpl::list<>;
```
why doesn't Burgers need this?

moxcodes replied:
Burgers has `has_primitive_and_conserved_vars = false`, so it doesn't take the constexpr branch that makes a distinction between the conservative and non-conservative variables in `SetVariables`.

I could add it to Burgers anyways, though, if it seems like the interface should be consistent among systems that are in flux conservative form.

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt:
```diff
@@ -0,0 +1,83 @@
+  Limiters
+  LinearOperators
+  MathFunctions
+  RelativisticEulerSolutions
```
are these supposed to be alphabetical?

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt:
```diff
@@ -0,0 +1,83 @@
+  Time
+  Utilities
+  ValenciaDivClean
+  GhValenciaDivClean
```
alphabet?

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
```diff
@@ -0,0 +1,493 @@
+          typename InitialData, typename... InterpolationTargetTags>
+struct GhValenciaDivCleanTemplateBase<
+    EvolutionMetavarsDerived<InitialData, InterpolationTargetTags...>>
+    : public virtual GhValenciaDivCleanDefaults {
```
why virtual inheritence?

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/EvolveGhValenciaDivClean.hpp:
```diff
@@ -0,0 +1,92 @@
+
+template <typename InitialData, typename... InterpolationTargetTags>
+struct EvolutionMetavars
+    : public virtual GhValenciaDivCleanDefaults,
```
why are you virtually inheriting from a class that is a base class of the other class you are inheriting from?

moxcodes replied:
I want all of the type aliases in `GhValenciaDivCleanDefaults` to be brought into the `EvolutionMetavars` directly, without having to specify `using ....` statements for each type alias. The type aliases are brought in automatically for public inheritance from the non-dependent `GhValenciaDivCleanDefaults`, but not from the template `GhValenciaDivCleanTemplateBase` that depends on the `EvolutionMetavars` template parameters -- this is ultimately why I have the two separate base classes.

Then, the virtual inheritance is because both the `EvolutionMetavars` and `GhValenciaDivCleanTemplateBase` inherit from `GhValenciaDivCleanDefaults` (the type aliases therein are also used in some definitions in `GhValenciaDivCleanTemplateBase`), so needs `virtual` to avoid inheritance ambiguities.

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/EvolveGhValenciaDivClean.hpp:
```diff
@@ -0,0 +1,92 @@
+      "Harmonic formulation\n"};
+};
+
+struct KerrHorizon {
```
why is this called KerrHorizon and why does it exist?

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
```diff
@@ -0,0 +1,493 @@
+      tmpl::append<typename system::primitive_variables_tag::tags_list,
+                   typename system::gh_system::variables_tag::tags_list>;
+  using ordered_list_of_primitive_recovery_schemes = tmpl::list<
+      grmhd::ValenciaDivClean::PrimitiveRecoverySchemes::NewmanHamlin,
```
might as well add the KastaunEtAl scheme

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/EvolveGhValenciaDivCleanWithHorizon.hpp:
```diff
@@ -0,0 +1,185 @@
+
+template <typename InitialData, typename... InterpolationTargetTags>
+struct EvolutionMetavars
+    : public virtual GhValenciaDivCleanDefaults,
```
why inherit again off of `GhValenciaDivCleanDefaults`?

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/EvolveGhValenciaDivCleanWithHorizon.hpp:
```diff
@@ -0,0 +1,185 @@
+                                         Triggers::time_triggers>>>;
+  };
+
+  using phase_changes =
```
why do you need these?  type aliases of public base classes should be visible

moxcodes replied:
Unfortunately, that doesn't happen for 'dependent' public base class templates, like in the CRTP trick.

kidder replied:
ah right...

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
```diff
@@ -138,6 +138,7 @@
 #include "PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp"
 #include "PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp"
 #include "PointwiseFunctions/AnalyticSolutions/RelativisticEuler/FishboneMoncriefDisk.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/RelativisticEuler/SmoothFlow.hpp"
```
should this be in the last commit?

kidder commented on src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:
```diff
@@ -0,0 +1,493 @@
+#include "PointwiseFunctions/AnalyticSolutions/GrMhd/BondiMichel.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/GrMhd/KomissarovShock.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/RelativisticEuler/FishboneMoncriefDisk.hpp"
```
should this include and the other 3 with horizons be in the next commit?

moxcodes commented:
Fixups posted, moved includes to appropriate commits. I haven't made any changes to the inheritance, pending further discussion.
Cheers!

nilsdeppe reviewed (commented):
fixup LGTM, I have no issues with the inheritance structure and the use of virtual inheritance as a result

kidder reviewed (commented):
you can squash; I can't think of anything to improve the inheritance

moxcodes commented:
fixups squashed in.
Cheers!

moxcodes commented:
Sorry, needed to rebase to bring in some factory creation changes.
Cheers!

kidder commented:
probably need another one since #3221 was just merged

moxcodes commented:
rebased again for good measure.
Cheers!

moxcodes commented:
Another rebase to fix conflicts.
Cheers!

kidder commented:
tests are failing because the output precision is not specified in the input files...

moxcodes commented:
pushed the alterations for the new input file parameters.
Cheers!
